ci: add `pytest` support on windows; change: add `meltano elt` guidance for windows users

visch opened the pull request:
Closes https://github.com/meltano/meltano/issues/5940

Old MR from Gitlab https://gitlab.com/meltano/meltano/-/merge_requests/2639/diffs

(Note this has been rebased since this MR)

Pipeline is running here https://gitlab.com/vischous/meltano/-/merge_requests/3/pipelines

visch committed:
Run pytest on Windows

visch committed:
Install pipx and poetry before tests run

visch committed:
Use proper Python versions for Meltano Tests on Windows

visch committed:
Keeping python3 / python somewhat consistent, added alias for gitlab ci job

visch committed:
Get these pipelines failing on PyTest as they should be

visch committed:
Point Windows users to use Meltano run instead of Meltano ELT.

visch committed:
CI should be referencing the correct ENV variable now for the Python version

visch committed:
CI used pytest job for the Windows pytest runner again

visch committed:
CI Should fail by running actual tests now, not pretty yet

visch committed:
100's more tests passing on windows, plus lint fixes

visch committed:
Ignored failing Windows Tests, and added mechanism for running all tests in a different directory

visch committed:
Update Windows ELT test to Legacy Tracker tracking like the rest

visch committed:
Add Windows Testing to Github Actions, and clean up the matrix a bit

visch commented on tests/fixtures/core.py:
```diff
@@ -471,7 +471,7 @@ def project(test_dir, project_init_service):
     Project.deactivate()
     os.chdir(test_dir)
-    shutil.rmtree(project.root)
+    # shutil.rmtree(project.root)
```
Should we do a full cleanup of all the pytest temp directories at the end of testing instead?

Also should we remove all the comments?

visch committed:
Add windows Specefic Github Action PyTest Run for Env Variables

visch committed:
Accidently readded test_repos.py durring Merge Conflict resolution, removed again

visch committed:
Added rmdir to test, PyTest should almost be passing now atleast on Linux

visch committed:
Fix Github action quotes messup

visch committed:
Remove hardcoded project name in Test

visch committed:
jobid changed names between making this PR originally, fixed typo

visch committed:
Merge branch 'main' into 2743-windows-elt-rebased

visch committed:
Make the UI look more clear as to which Systems test is running, Linux is nicer than Unix-Like I think

visch committed:
Errant import got in probably from an autoimport, causing issues with Python 3.7

visch committed:
Merge branch '2743-windows-elt-rebased' of github.com:AutoIDM/meltano into 2743-windows-elt-rebased

visch commented:
Still a few errors here https://gitlab.com/vischous/meltano/-/jobs/2576910300

edgarrmondragon commented on .github/workflows/test.yml:
```diff
@@ -11,10 +11,8 @@ jobs:
-        - { python-version: "3.7", os: "ubuntu-latest" }
-        - { python-version: "3.8", os: "ubuntu-latest" }
-        - { python-version: "3.9", os: "ubuntu-latest" }
+        python-version: ["3.7", "3.8", "3.9"]
```
@visch This notation is easier to understand and it's ok as long as the combination is a cartesian product

visch replied:
It does the cartesian product! :D

visch commented on .gitlab/ci/global.gitlab-ci.yml:
```diff
@@ -29,3 +29,8 @@
     matrix:
       - IMAGE_TAG_SUFFIX: ["", "-python3.7", "-python3.8", "-python3.9"]
+
+.windows:defaults:
```
Not removing this only because of https://github.com/meltano/meltano/pull/5967/files

visch committed:
Remove Windows gitlab pytest tests

visch commented:
> Still a few errors here https://gitlab.com/vischous/meltano/-/jobs/2576910300

FAILED tests/meltano/core/test_plugin_install_service.py::TestPluginInstallService::test_install_all
FAILED tests/meltano/core/test_state_service.py::TestStateService::test_set_state

These work properly on my Windows machine, just not on Windows CI at gitlab for some reason.  

They are passing https://github.com/meltano/meltano/pull/6155/checks though? Which is curious, maybe that's enough to ignore them?

I'm removed the Gitlab Windows Checks and we'll just rely on the Windows Github checks.

visch committed:
Merge branch 'main' into 2743-windows-elt-rebased

visch committed:
New Test doesn't work on Windows, ignoring the test for now

visch commented on tests/meltano/core/tracking/test_exception.py:
```diff
@@ -97,6 +99,10 @@ def test_simple_exception_context():
         assert key in tb_data[0]
 
 
+@pytest.mark.skipif(
```
This is pretty important as it's an example of a new Test was was written that doesn't pass on Windows. I went ahead and ignored this test. We'll probably want to at a minimum add to a guide that if a pass doesn't test on Windows you'll want to "ignore" the test, although that sounds terrible at least it won't impact development velocity as badly.

Punting these questions to https://github.com/meltano/meltano/issues/3444 is ok, but it does mean that we are currently telling everyone to just ignore all the Windows failures until we have a better solution here https://github.com/meltano/meltano/issues/3444 which will have to be some form of either always ignore the error, fix the error, or bring someone in to help fix the error.

visch replied:
After chatting with @aaronsteers  it sounds like this is fine. I'll leave this open and let @aaronsteers  resolve just in case :D

WillDaSilva replied:
I believe I've fixed the Windows incompatibility in my branch for Python 3.10 support. I expect this PR will be merged first, so it'll be easy for me to verify that.

aaronsteers replied:
@visch and @WillDaSilva - Resolving this. Yes, makes sense. 👍

visch committed:
Gitlab -> Github issue links

visch commented:
@edgarrmondragon  actually ready for a review now, not a fake out this time!

edgarrmondragon commented on .github/workflows/test.yml:
```diff
@@ -52,12 +50,18 @@ jobs:
 
     - name: Upload coverage data
-      if: always() && (matrix.python-version == '3.9')
+      if: always() && (matrix.python-version == '3.9') && (matrix.os == 'ubuntu-latest')
```
@aaronsteers any reason we don't want to combine coverage from multiple Python versions? If not, I guess there's also no reason not to combine coverage from multiple OSs. Wdyt?

> Often test suites are run under different conditions, for example, with different versions of Python, or dependencies, or on different operating systems. In these cases, you can collect coverage data for each test run, and then combine all the separate data files into one combined file for reporting.

https://coverage.readthedocs.io/en/latest/cmd.html#cmd-combine

aaronsteers replied:
@edgarrmondragon and @visch - No reasons I'm aware to collect and combine coverage. This snippet was written before we had a combine step (or before I understood it. 😄 )

We can open this as a separate issue though rather than delay to PR.

edgarrmondragon commented on tests/fixtures/cli.py:
```diff
@@ -39,5 +39,5 @@ def project_files_cli(test_dir, compatible_copy_tree):
     Project.deactivate()
     os.chdir(test_dir)
-    shutil.rmtree(project.root)
+    # shutil.rmtree(project.root)
```
@visch do you know if these files would linger indefinitely in a dev's system when running tests locally?

edgarrmondragon commented on .github/workflows/test.yml:
```diff
@@ -52,12 +50,18 @@ jobs:
         poetry install
 
-    - name: Run pytest
+    - name: Run pytest linux
+      if: (matrix.os != 'windows-latest')
       run: |
         poetry run coverage run --parallel -m pytest -m "$PYTEST_MARKERS"
+    
+    - name: Run pytest windows
+      if: (matrix.os == 'windows-latest')
+      run: |
+        poetry run coverage run --parallel -m pytest -m $Env:PYTEST_MARKERS
```
@visch Since the pytest markers string doesn't really need to be an env var, wdyt of adding `pytest_markers: ["not concurrent"]` to the matrix and using a single step with

```yaml
    - name: Run pytest
      run: |
        poetry run coverage run --parallel -m pytest -m "${{ matrix.pytest_markers }}"
```

visch replied:
Fixed, did it with Env variables looks like this is how do you "expansion" with github actions.

edgarrmondragon replied:
@visch Ah, even better. Thanks!

visch committed:
Combine windows and linux tests to simply workflow

visch committed:
Merge branch '2743-windows-elt-rebased' of github.com:AutoIDM/meltano into 2743-windows-elt-rebased

visch committed:
Merge branch 'main' into 2743-windows-elt-rebased

visch committed:
Merge branch 'main' into 2743-windows-elt-rebased

visch commented:
From the latest Postgres add to github actions failure

```
https://github.community/t/how-to-run-services-on-windows-or-macos/16565
```

We could just roll with gitlab stuff for now, but I"d guess we want to spend the time to fix this. 

My gut "want" is some kind of inheritance  where the Windows tests don't use Postgres but can (somehow) still run the same Steps as the current github actions. Not sure how exactly but there may be an option. Worst case we make a `test-windows.yml`

visch committed:
Remove commented out rmdir lines

visch committed:
Remove Windows From Postgres Tests as services do not work in Windows/macOS containers

visch commented:
> From the latest Postgres add to github actions failure
> 
> ```
> https://github.community/t/how-to-run-services-on-windows-or-macos/16565
> ```
> 
> We could just roll with gitlab stuff for now, but I"d guess we want to spend the time to fix this.
> 
> My gut "want" is some kind of inheritance where the Windows tests don't use Postgres but can (somehow) still run the same Steps as the current github actions. Not sure how exactly but there may be an option. Worst case we make a `test-windows.yml`

@edgarrmondragon  I pushed up a change removing Windows from the test with PostGres and adding a new `test-nopostgres.yml` file that works without it.  I know this is a bad solution (But it works), what do you think about installing the Postgres container manually (Not using a service, using a runner instead) so that we can workaround not having conditional services https://github.community/t/conditional-services-in-a-job/135301

edgarrmondragon commented on .github/workflows/test-nopostgres.yml:
```diff
@@ -0,0 +1,107 @@
+        name: coverage-data
+        path: ".coverage.*"
+
+  coverage:
```
@visch see https://github.com/meltano/meltano/pull/6268

visch committed:
Merge branch 'main' into 2743-windows-elt-rebased

visch committed:
Add Windows back to test.yml as github action should now work on Windows!

visch committed:
Windows-latest

visch committed:
Only test Windows with sqllite

visch commented:
> Windows Postgres failures
> 
> ```
> Run docker run -d -p "5432:5432" -e "POSTGRES_PASSWORD=postgres" --name postgres --health-cmd "pg_isready -d postgres -U postgres" --health-interval [1](https://github.com/meltano/meltano/runs/7042154323?check_suite_focus=true#step:7:1)0s --health-timeout 5s --health-retries 5 postgres:11
> Unable to find image 'postgres:11' locally
> 11: Pulling from library/postgres
> docker: no matching manifest for windows/amd64 10.0.2034[8](https://github.com/meltano/meltano/runs/7042154323?check_suite_focus=true#step:7:9) in the manifest list entries.
> See 'docker run --help'.
> Error: Process completed with exit code 1.
> ```
> 
> Few ways to handles this
> 
> 1. Ignore installing postgres on Windows, add if's in the GH actions
> 2. Barring some postgres image that works on Windows we could use WSL - [Add Windows Subsystem for Linux to Windows actions/virtual-environments#50](https://github.com/actions/virtual-environments/issues/50) which would mean:
>    
>    1. Install WSL
>    2. Install Docker on WSL2 (🤞 that this would work as it runs on HyperV)
>    3. Install Postgres via Docker on WSL2
>    4. Setup networking so that you can route to postgres, on my Windows box I do static IP mapping to the interface HyperV uses but there may be an easier way here
>    5. Write up some dev docs on how to use Docker Containers and interact with them via CI for folks that want to add mssql, mysql, and others ([Add broader system db support #6167](https://github.com/meltano/meltano/pull/6167))
> 3. Install postgres directly on windows (works pretty well and is simple, you have to worry a bit about dependencies on the local box impacting pytest in some unexpected way)
> 4. ~https://stackoverflow.com/questions/48066994/docker-no-matching-manifest-for-windows-amd64-in-the-manifest-list-entries - going to try this first.~ Nope :( and this might crush our wsl2 ideas) [Add Windows Agent with Docker in Linux container mode actions/virtual-environments#252 (comment)](https://github.com/actions/virtual-environments/issues/252#issuecomment-610370237)

Submitted https://github.com/meltano/meltano/issues/6281 instead of trying to fix this here.

WillDaSilva reviewed (changes requested):
The tests that fail on Windows, but shouldn't, should use `pytest.xfail` instead of `pytest.skipif`. The Pytest docs for xfail have an example for conditionally xfailing on Windows: https://docs.pytest.org/en/7.1.x/how-to/skipping.html#condition-parameter

Skipping is a neutral signal. It indicates that a test could not be run, but doesn't tell us whether anything is wrong with that or not. An xfail on the other hand tells us that there is a problem, but it's expected, and we ought to have an issue open to address it.

edgarrmondragon commented on .github/workflows/test.yml:
```diff
@@ -15,6 +15,9 @@ jobs:
         - { python-version: "3.7", os: "ubuntu-latest", backend-db: sqlite }
         - { python-version: "3.8", os: "ubuntu-latest", backend-db: sqlite }
         - { python-version: "3.9", os: "ubuntu-latest", backend-db: sqlite }
+        - { python-version: "3.7", os: "windows-latest", backend-db: sqlite }
+        - { python-version: "3.8", os: "windows-latest", backend-db: sqlite }
+        - { python-version: "3.9", os: "windows-latest", backend-db: sqlite }
```
👍 hopefully we can start testing Windows + Postgres at some point

visch replied:
https://github.com/meltano/meltano/issues/6281 :D yes, I'll add this issue link inline as well

visch commented:
> The tests that fail on Windows, but shouldn't, should use `pytest.xfail` instead of `pytest.skipif`. The Pytest docs for xfail have an example for conditionally xfailing on Windows: https://docs.pytest.org/en/7.1.x/how-to/skipping.html#condition-parameter
> 
> Skipping is a neutral signal. It indicates that a test could not be run, but doesn't tell us whether anything is wrong with that or not. An xfail on the other hand tells us that there is a problem, but it's expected, and we ought to have an issue open to address it.

I think you're right, but some of the tests on Windows lead to the job hanging the entire pytest process. I didn't take the time to figure out which tests fail and which ones hang I could do that here if you all want me to?

How do you feel about punting that to https://github.com/meltano/meltano/issues/3444 I added a comment there? I think that issue is going to end up being a bunch of sub issues addressing individual failures, and one of those should be switching to xfail for everything.

visch committed:
Inline comment pointing to why Windows isn't testing other dbs yet

WillDaSilva commented:
> How do you feel about punting that to https://github.com/meltano/meltano/issues/3444 I added a comment there? I think that issue is going to end up being a bunch of sub issues addressing individual failures, and one of those should be switching to xfail for everything.

Works for me, though an alternative that lets us use xfail and avoid the having the process hang would be adding something like this at the start of each of these failing tests:

```python
def test_something(...):
    if platform.system() == "Windows":
        pytest.xfail("reason...")
    # The original body of the test function continues here
```

visch committed:
Switched to xfail from skipif

visch committed:
Switched to xfail from skipif

visch committed:
Line feed correction

visch commented on tests/meltano/core/logging/test_output_logger.py:
```diff
@@ -186,7 +211,8 @@ def test_logging_exception(self, log, subject, redirect_handler):
         # make sure it let the exception through
-        assert exc.value is exception
+        # All code below here in this test cannot be reached
+        assert exc.value is exception  # noqa: WPS441
```
Note I didn't fix all the linting issues I started to and gave up on it

visch commented:
> > How do you feel about punting that to #3444 I added a comment there? I think that issue is going to end up being a bunch of sub issues addressing individual failures, and one of those should be switching to xfail for everything.
> 
> Works for me, though an alternative that lets us use xfail and avoid the having the process hang would be adding something like this at the start of each of these failing tests:
> 
> ```python
> def test_something(...):
>     if platform.system() == "Windows":
>         pytest.xfail("reason...")
>     # The original body of the test function continues here
> ```



> > How do you feel about punting that to #3444 I added a comment there? I think that issue is going to end up being a bunch of sub issues addressing individual failures, and one of those should be switching to xfail for everything.
> 
> Works for me, though an alternative that lets us use xfail and avoid the having the process hang would be adding something like this at the start of each of these failing tests:
> 
> ```python
> def test_something(...):
>     if platform.system() == "Windows":
>         pytest.xfail("reason...")
>     # The original body of the test function continues here
> ```

Went ahead and added xfail instead of skip!

visch committed:
Issue number typo

WillDaSilva commented on tests/meltano/cli/test_elt.py:
```diff
@@ -219,6 +220,39 @@ def dbt_process(process_mock_factory, dbt):
+
+@pytest.mark.skipif(
+    platform.system() == "Windows",
+    reason="ELT doesn't work for Windows, don't run these tests on Windows as they fail in horrible ways.",
```
```suggestion
    reason="ELT is not supported on Windows",
```

WillDaSilva commented on tests/meltano/cli/test_elt.py:
```diff
@@ -1012,6 +1046,10 @@ def test_dump_loader_config(
 
+@pytest.mark.skipif(
+    platform.system() == "Windows",
+    reason="ELT doesn't work for Windows, don't run these tests on Windows as they fail in horrible ways.",
```
```suggestion
    reason="ELT is not supported on Windows",
```

WillDaSilva commented on tests/meltano/cli/test_elt.py:
```diff
@@ -1179,6 +1217,10 @@ def test_elt_transform_run_dbt_failure(
 
+@pytest.mark.skipif(
+    platform.system() == "Windows",
+    reason="ELT doesn't work for Windows, don't run these tests on Windows as they fail in horrible ways.",
```
```suggestion
    reason="ELT is not supported on Windows",
```

WillDaSilva reviewed (approved):
Thanks for getting this done @visch! I'm glad we'll be able to make these Windows tests available in CI in time for me to run the Python 3.10 changes against them.

visch committed:
Merge branch 'main' into 2743-windows-elt-rebased

visch committed:
New Test Tracker tests fail on Windows as well

visch committed:
Xfail isn't enough to prevent a failure for these tests

visch committed:
Apply suggestions from code review

Better skip reason

Co-authored-by: Will Da Silva <will@willdasilva.xyz>

visch committed:
PROJECT_NAME is not referenced anywhere

visch committed:
Merge branch '2743-windows-elt-rebased' of github.com:AutoIDM/meltano into 2743-windows-elt-rebased

visch commented:
Code Coverage stuff isn't quite accurate here (Or maybe I misunderstand) as we are testing these things but it's dependent on our github actions. We could dive into this as a part of this PR as well? (this is new as well!)

visch committed:
Merge branch 'main' into 2743-windows-elt-rebased

edgarrmondragon commented:
> I guess my question here is do we need to fix code coverage here or not?

@visch not at all. It's currently only a guideline and not a strict requirement.

visch committed:
Merge conflicts + auto changes from black and other precommit tools

visch committed:
Merge conflicts fixed (round 2 for today :D )

visch committed:
New test fails on Windows, added xfail

WillDaSilva commented:
@visch The new tests failures are my bad: #6306 

Sorry about that. I'll have it sorted out as soon as possible.
